idaes/core/solvers/petsc.py

```diff
@@ -0,0 +1,688 @@
+            delete_on_read (bool): if true delete trajectory data after reading
+            unscale (Block): if not None this is a block to read scale factors
+                to unscale the trajectory
+        """
+        if PetscBinaryIOTrajectory is None:
+            raise RuntimeError("PetscBinaryIOTrajectory could not be imported")
+        if pth is not None:
+            stub = os.path.join(pth, stub)
+            vis_dir = os.path.join(pth, vis_dir)
+        if stub is not None:
+            self.stub = stub
+            self.vis_dir = vis_dir
+            self.path = pth
+            self.unscale = unscale
+            self._read()
+            if delete_on_read:
+                self.delete_files()
+            if unscale is not None:
+                self._unscale(unscale)
+        elif vecs is not None:
+            self.vecs = vecs
+            self.time = vecs["_time"]
+            self.vars = list(vecs.keys())
+        elif json is not None:
+            self.from_json(json)
+        else:
+            raise RuntimeError("To read trajectory, either provide stub, vecs, or json")
+
+    def _read(self):
+        with open(f'{self.stub}.col') as f:
+            names = list(map(str.strip, f.readlines()))
+        with open(f'{self.stub}.typ') as f:
+            typ = list(map(int,f.readlines()))
+        self.vars = [name for i, name in enumerate(names) if typ[i] in [0,1]]
+        (t, v, names) = PetscBinaryIOTrajectory.ReadTrajectory(
+            "Visualization-data"
+        )
+        self.time = t
+        self.vecs_by_time = v
+        self.vecs = dict.fromkeys(self.vars, None)
+        for k in self.vecs.keys():
+            self.vecs[k] = [0]*len(self.time)
+        self.vecs["_time"] = list(self.time)
+        for i, v in enumerate(self.vars):
+            for j, vt in enumerate(self.vecs_by_time):
+                self.vecs[v][j] = vt[i]
+
+    def _var_name_list(self, vars):
+        if isinstance(vars, str):
+            vars = [vars]
+        if hasattr(vars, "ctype"):
+            if issubclass(vars.ctype, pyo.Var):
+                vars = [vars]
+        vars = list(map(str, vars))
+        for name in vars:
+            if name not in self.vars:
+                raise RuntimeError(f"Variable {name} not found.")
+        return vars
+
+    def get_vec(self, var):
+        """Return the time vector for var.
+
+        Args:
+            var (str or Var): Variable to get vector for.
+
+        Retruns:
+            (list): time vector
+
+        """
+        var = str(var)
+        return self.vecs[var]
+
+    def get_dt(self):
+        """Get a list of time steps
+
+        Args:
+            None
+
+        Returns:
+            (list)
+        """
+        dt = [None]*(len(self.time) - 1)
+        for i in range(len(self.time)-1):
+            dt[i] = self.time[i + 1] - self.time[i]
+        return dt
+
+    def interpolate_vecs(self, times):
```

Comment:
There is an easy way to do this using Numpy's interpolation function which I expect would greatly improve the performance of this method. See: https://github.com/Pyomo/pyomo/blob/main/pyomo/dae/simulator.py#L1058

Reply:
My thinking on this was if I had a lot of time point and a lot of variables, I could save time searching by assuming the times were in order then reusing the index and weights for all the variables.  I'm sure you are right.  Numpy should be a lot faster.  I'll take a closer look at this see if I can find the most efficient way to do it.

Reply:
Updated to Numpy (which is likely faster and definitely cleaner).  I still feel there is probably a better way than what I have if you share part of the computation between between all the variables, but it's probably not worth worrying about.  If slowness is a problem for some super large problem we can probably improve.